packages/playwright-crawler/src/internals/rendering-type-prediction.ts

```diff
@@ -0,0 +1,95 @@
+import LogisticRegression from 'ml-logistic-regression';
+import { Matrix } from 'ml-matrix';
+import stringComparison from 'string-comparison';
+
+export type RenderingType = 'clientOnly' | 'static'
+
+type URLComponents = string[];
+
+const urlComponents = (url: URL): URLComponents => {
+    return [url.hostname, ...url.pathname.split('/')];
+};
+
+const calculateUrlSimilarity = (a: URLComponents, b: URLComponents): number | undefined => {
+    const values: number[] = [];
+
+    if (a[0] !== b[0]) {
+        return 0;
+    }
+
+    for (let i = 1; i < Math.max(a.length, b.length); i++) {
+        values.push(stringComparison.jaroWinkler.similarity(a[i] ?? '', b[i] ?? '') > 0.8 ? 1 : 0);
+    }
+
+    return sum(values) / Math.max(a.length, b.length);
+};
+
+const sum = (values: number[]) => values.reduce((acc, value) => acc + value);
+const mean = (values: number[]) => (values.length > 0 ? sum(values) / values.length : undefined);
+
+type FeatureVector = [staticResultsSimilarity: number, clientOnlyResultsSimilarity: number];
+
+export class RenderingTypePredictor {
+    private renderingTypeDetectionResults = new Map<RenderingType, URLComponents[]>();
+    private detectionRatio: number;
+    private logreg: LogisticRegression;
+
+    constructor({ detectionRatio }: { detectionRatio: number }) {
+        this.detectionRatio = detectionRatio;
+        this.logreg = new LogisticRegression({ numSteps: 1000, learningRate: 5e-2 });
```

Comment:
World's smallest nit, but why not just write 5e-2 as 0.05? 👀 Or maybe it's just me not being used to scientific notations yet 😄

Reply:
Because this is science, of course! Honestly, it's copypasted code.